### DISCUSSION TITLE
Adding `executor_config` slow down the pods initialization on Kubernetes ?

### DISCUSSION THREAD

**Discussion by KYPcode** [apache/airflow#35700] [Q&A] [answered]
Hello,

First thanks for this awesome tool.

I'd like to add limit and per-request resources to my Airflow Task pods (on kubernetes) with the python @task operator.
I therefore chose the option of adding the `executor_config` parameter as follows to the DAG arguments (https://github.com/apache/airflow/discussions/20149). However, when I do this, I my pods take twice as long to initialize. Is there a reason to that? Am I missing a best practice here?

```
'executor_config':{
    "KubernetesExecutor": {
        "request_cpu": "250m",
        "limit_cpu": "400m",
        "request_memory": "150Mi",
        "limit_memory": "300Mi"
    }
}
```

**Comment by potiuk** [comment, marked as answer]
There are two options:

* tt just takes your K8s more time to create  such a POD (maybe because reserving resources takes time to find the right node and to actually reserve the resources. you can test it by attempting to create such pod manually and experimenting with various parameters - generally experimenting with various parameters and choosing the right settings i part of deployment management task and somethign that only you can do for your particular deployment

* it might be tha that by limiting memory you are stressing the worker to take more time to initialize - you are limiting it to 0.5 CPI and 300 M memory - and this is how much resources the interpreter has to run it's task. Generally if you have CPU or memory constrained process then it will, well, limit the resources it will use. Which migh turn into a slower startup time, This is obvious and expected. If you have a process that calculates somethin in a CPU intensive loop and give it half-CPU, it will take 2x longer to complete. Same with memory - if your task will require more memory than 300M and your K8s (modern  kubernetes do) allow it to use swap disk memory - it will start using swap and slow down multiple times sometimes.

Generally speaking - it's your task as deployment manager to setup monitoring and observe resource usage and fine-tune and adjust your resource limits to get the most of it - if you see your tasks are maxing out CPU usage - you should probably increase it, same with all the other resources. Sometimes it means that you just need to increase resources for your deployment if you want to optimise speed. The usual drill - nothing 'airflow special' there.

There is no "by the book" solution here - it all depends on what your tasks do , what resources are made available to them and your fine tuning should optimise for what you want - usually more resources means faster execution - but it's your decision whether you want to optimise for speed or resources and monitoring that you should setup should help you with deciding it.

Airflow does not have "resource optimising" built in - this is something extremely complex (especailly that you need to know what optimisation goals you have) and it's left for discretion of deployment managers to decide.

**Reply by KYPcode** [reply]
Thanks a lot for this very complete answer @potiuk, I'm going to look into it.
